temporal-sdk/src/main/java/io/temporal/client/WorkflowClient.java

```diff
@@ -237,7 +239,29 @@ WorkflowStub newUntypedWorkflowStub(
    */
   WorkflowExecution signalWithStart(BatchRequest signalWithStartBatch);
 
-  WorkflowServiceStubs getWorkflowServiceStubs();
+  /**
+   * A wrapper around {WorkflowServiceStub#listWorkflowExecutions(ListWorkflowExecutionsRequest)}
+   *
+   * <p>Note: This method uses 1000 as the underlying page size. To customize the page size, use
+   * {@link #listWorkflowExecutions(String, int)}
+   *
+   * @param query Temporal Visibility Query, for syntax see <a
+   *     href="https://docs.temporal.io/visibility#list-filter">Visibility docs</>
+   * @return iterable that performs remote pagination under the hood
+   */
+  default Iterable<WorkflowExecutionMetadata> listWorkflowExecutions(String query) {
+    return listWorkflowExecutions(query, 1000);
+  }
+
+  /**
+   * A wrapper around {WorkflowServiceStub#listWorkflowExecutions(ListWorkflowExecutionsRequest)}
+   *
+   * @param query Temporal Visibility Query, for syntax see <a
+   *     href="https://docs.temporal.io/visibility#list-filter">Visibility docs</>
+   * @param pageSize a page size to use for the underlying remote pagination
+   * @return iterable that performs remote pagination under the hood
+   */
+  Iterable<WorkflowExecutionMetadata> listWorkflowExecutions(String query, int pageSize);
```

Comment:
This seems right to me, but I just want to confirm it's still normal in modern Java to return iterables/iterators and have the user convert to streams if they want as opposed to returning a stream for them?

Reply:
I'm a little reserved about using Stream in public APIs of libraries if stream manipulation is not the sole purpose of these APIs. If the producer and consumer of the Stream are in the same codebase - sure, why not, it's handy.

Otherwise, Stream instance carries a bunch of configurations with it that the producer would define if Stream is in the method definition. Specifically, Spliterator characteristics and if the Stream is parallel or not. These configurations are often an implementation detail of a CONSUMER and important for a CONSUMER, not a producer. Exposing Stream, we make consumers who care about these parameters of the Stream to

guess and expect that parameters are correct
assert that the returned Stream has expected characteristics
or recreate the Stream with the correct characteristics.
I think it's a less error-prone experience to return an unambiguous Iterable and let the consumer create a Stream with the parameters that the consumer expects / relies on.

But I may see how users may no want to do that and expect a simpler higher-level API and if you have a strong opinion - I can rework on returning a Stream.

Reply:
I don't have a strong opinion here. The only thing I might suggest is returning your own `WorkflowExecutionMetadataIterable` here with a `asStream()` on it that knows we don't support parallel. But I think we can add that later and still be considered compatible (I figure we're allowed to raise to higher interface in covariant returns even though it technically breaks explicit implementers of the interface). The real benefit for me for streams was the ability to map and collect and such. From that issue, I was hoping something nice like:

```java
public void runBeforeDeployment(WorkflowClient client, Worker worker, String workflowType) {
  client
    .list("WorkflowType = '" + workflowType + "'")
    .map(client::fetchHistory)
    .parallel()
    .map(worker::replayWorkflowExecution);
}
```

Could exist. Here it adds an extra step, but not that big of a deal.

Reply:
> asStream() on it that knows we don't support parallel

That's exactly my point for not doing this. Users should define if they want a parallel or not parallel stream. There is nothing wrong to create a parallel stream from this iterator.

Reply:
That works for me

Reply:
Reconsidered it and will end up exposing the Stream here. The main reason is that correct support of Iterable would mean that iterator() should create a new iterator if called several times and it becomes a bit too much for this method.